### DISCUSSION TITLE
Course, Chapter 3, entities in 7) Complex components

### DISCUSSION THREAD

**Discussion by oxidian24** [explosion/spaCy#12191] [Help: Best practices]
Hi,
I just wanted to ask you what's the point of creating the entities in the chapter 3, point 7. I mean, why this example uses this approach, why not just to list the span directly? Does this lesson mean that I can define any data I want to be collected in the entities?

```
@Language.component("animal_component")
def animal_component_function(doc):
    matches = matcher(doc)
    spans = [Span(doc, start, end, label="ANIMAL") for match_id, start, end in matches]
    doc.ents = spans
    return doc
```

I just dont understand why this needs to be collected in the span and with the ANIMAL label, why not just using the doc[x:y] approach?

**Comment by danieldk** [comment]
> Does this lesson mean that I can define any data I want to be collected in the entities? [...] Does this lesson mean that I can define any data I want to be collected in the entities?

You can extract entities however you would like and assign them to `doc.ents`, however the list or tuple that you assign to `doc.ents` [must consist of spans](https://spacy.io/api/doc#ents).

> I just dont understand why this needs to be collected in the span and with the ANIMAL label, why not just using the doc[x:y] approach?

`doc[start:end]` would create a span without a label and spans without labels are ignored when assigning to `doc.ents`. So a solution with doc slicing would look something like this:

```
@Language.component("animal_component")
def animal_component_function(doc):
    matches = matcher(doc)
    spans = []
    for match_id, start, end in matches:
        span = doc[start:end]
        span.label_ = "ANIMAL"
        spans.append(span)
    doc.ents = spans
    return doc
```

which is less compact than the solution in the course.
